docker/docker_image_dest.go

```diff
@@ -332,21 +332,31 @@ func (d *dockerImageDestination) TryReusingBlobWithOptions(ctx context.Context,
 	// Then try reusing blobs from other locations.
 	candidates := options.Cache.CandidateLocations2(d.ref.Transport(), bicTransportScope(d.ref), info.Digest, options.CanSubstitute)
 	for _, candidate := range candidates {
-		candidateRepo, err := parseBICLocationReference(candidate.Location)
-		if err != nil {
-			logrus.Debugf("Error parsing BlobInfoCache location reference: %s", err)
-			continue
+		var candidateRepo reference.Named
+		if !candidate.UnknownLocation {
+			candidateRepo, err = parseBICLocationReference(candidate.Location)
+			if err != nil {
+				logrus.Debugf("Error parsing BlobInfoCache location reference: %s", err)
+				continue
+			}
 		}
 		if candidate.CompressorName != blobinfocache.Uncompressed {
-			logrus.Debugf("Trying to reuse cached location %s compressed with %s in %s", candidate.Digest.String(), candidate.CompressorName, candidateRepo.Name())
+			logrus.Debugf("Trying to reuse cached location %s compressed with %s", candidate.Digest.String(), candidate.CompressorName)
 		} else {
-			logrus.Debugf("Trying to reuse cached location %s with no compression in %s", candidate.Digest.String(), candidateRepo.Name())
+			logrus.Debugf("Trying to reuse cached location %s with no compression", candidate.Digest.String())
 		}
 
 		// Sanity checks:
-		if reference.Domain(candidateRepo) != reference.Domain(d.ref.ref) {
-			logrus.Debugf("... Internal error: domain %s does not match destination %s", reference.Domain(candidateRepo), reference.Domain(d.ref.ref))
-			continue
+		if candidate.UnknownLocation || reference.Domain(candidateRepo) != reference.Domain(d.ref.ref) {
+			if !candidate.UnknownLocation {
+				logrus.Debugf("... Internal error: domain %s does not match destination %s", reference.Domain(candidateRepo), reference.Domain(d.ref.ref))
+				continue
+			}
+			// Found following blob in different registry but
+			// we need to check blob presence against the registry
+			// where we are planning to push, hence switch back the
+			// candidate repo to the one where we are planning to push.
+			candidateRepo = d.ref.ref
```

Comment:
```suggestion
			candidateRepo = reference.TrimNamed(d.ref.ref)
```
per one of the other threads.